File: backend/domain/eduproject/entity/project.go
/*
 * Copyright 2025 coze-dev Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package entity

import (
	"time"
)

// StudentProject 学生项目实体
type StudentProject struct {
	ID           int64     `json:"id" gorm:"primaryKey"`
	SpaceID      int64     `json:"space_id" gorm:"type:bigint unsigned;not null;index"`   // 所属空间ID
	StudentID    int64     `json:"student_id" gorm:"index;not null"`      // 关联user表
	ScriptID     int64     `json:"script_id" gorm:"index;not null"`       // 关联edu_scripts表
	Title        string    `json:"title" gorm:"type:varchar(100);not null"`
	CurrentStage int8      `json:"current_stage" gorm:"type:tinyint;not null;default:1"`
	Status       string    `json:"status" gorm:"type:varchar(20);not null;default:'in_progress'"` // in_progress/completed/abandoned
	Progress     int8      `json:"progress" gorm:"type:tinyint;not null;default:0"`               // 进度百分比
	StartedAt    time.Time `json:"started_at" gorm:"autoCreateTime"`
	CompletedAt  *time.Time `json:"completed_at,omitempty"`
	CreatedAt    time.Time `json:"created_at" gorm:"autoCreateTime"`
	UpdatedAt    time.Time `json:"updated_at" gorm:"autoUpdateTime"`
}

// TableName 指定表名
func (StudentProject) TableName() string {
	return "edu_student_projects"
}

// ProjectStage 项目阶段记录实体
type ProjectStage struct {
	ID             int64      `json:"id" gorm:"primaryKey"`
	ProjectID      int64      `json:"project_id" gorm:"index;not null"`
	StageNumber    int8       `json:"stage_number" gorm:"type:tinyint;not null"`
	StageName      string     `json:"stage_name" gorm:"type:varchar(50);not null"`
	Status         string     `json:"status" gorm:"type:varchar(20);not null;default:'pending'"` // pending/in_progress/completed
	OutputDocument string     `json:"output_document" gorm:"type:longtext"`                      // Markdown文档
	StartedAt      *time.Time `json:"started_at,omitempty"`
	CompletedAt    *time.Time `json:"completed_at,omitempty"`
	CreatedAt      time.Time  `json:"created_at" gorm:"autoCreateTime"`
}

// TableName 指定表名
func (ProjectStage) TableName() string {
	return "edu_project_stages"
}

// ProjectStatus 项目状态常量
const (
	ProjectStatusInProgress = "in_progress"
	ProjectStatusCompleted  = "completed"
	ProjectStatusAbandoned  = "abandoned"
)

// StageStatus 阶段状态常量
const (
	StageStatusPending    = "pending"
	StageStatusInProgress = "in_progress"
	StageStatusCompleted  = "completed"
)

File: frontend/packages/edu-learning/src/api/project.ts
/*
 * Copyright 2025 coze-dev Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import apiClient, { BaseResponse, PaginatedResponse } from './client';
import type {
  ProjectType,
  ProjectStatus,
  StudentProject,
  ProjectStage,
  EvaluationInfo,
} from '@coze-edu/common/types';

// ========== 请求类型 ==========

export interface CreateProjectRequest {
  space_id: number;
  project_type: ProjectType;
  source_id: number;
  title: string;
  description?: string;
  class_id?: number;
  assignment_id?: number;
}

export interface ListProjectsRequest {
  space_id?: number;
  project_type?: ProjectType;
  status?: ProjectStatus;
  keyword?: string;
  page?: number;
  page_size?: number;
}

export interface UpdateStageOutputRequest {
  stage_id: number;
  content: string;
}

export interface CompleteStageRequest {
  project_id: number;
  stage_order: number;
  output_content: string;
}

// ========== 响应类型 ==========

export interface ProjectDetail extends StudentProject {
  stages?: ProjectStage[];
  evaluations?: EvaluationInfo[];
}

export interface StageCompletionData {
  stage_id: number;
  score: number;
  feedback: string;
  next_stage?: number;
  evaluation?: EvaluationInfo;
}

// ========== API 函数 ==========

/**
 * 创建学习项目
 */
export const createProject = async (
  spaceId: number,
  request: CreateProjectRequest
): Promise<StudentProject> => {
  const response: BaseResponse<StudentProject> = await apiClient.post(
    `/api/space/${spaceId}/edu/projects`,
    request
  );
  return response.data!;
};

/**
 * 获取项目详情
 */
export const getProject = async (
  spaceId: number,
  projectId: number
): Promise<ProjectDetail> => {
  const response: BaseResponse<ProjectDetail> = await apiClient.get(
    `/api/space/${spaceId}/edu/projects/${projectId}`
  );
  return response.data!;
};

/**
 * 获取项目列表
 */
export const listProjects = async (
  spaceId: number,
  request?: ListProjectsRequest
): Promise<PaginatedResponse<StudentProject>> => {
  const response: BaseResponse<PaginatedResponse<StudentProject>> = await apiClient.get(
    `/api/space/${spaceId}/edu/projects`,
    { params: request }
  );
  return response.data!;
};

/**
 * 获取我的项目列表
 */
export const getMyProjects = async (spaceId: number): Promise<StudentProject[]> => {
  const response: BaseResponse<StudentProject[]> = await apiClient.get(
    `/api/space/${spaceId}/edu/projects/my`
  );
  return response.data!;
};

/**
 * 更新阶段产出内容
 */
export const updateStageOutput = async (
  spaceId: number,
  request: UpdateStageOutputRequest
): Promise<void> => {
  await apiClient.put(`/api/space/${spaceId}/edu/stages/output`, request);
};

/**
 * 完成当前阶段
 */
export const completeStage = async (
  spaceId: number,
  request: CompleteStageRequest
): Promise<StageCompletionData> => {
  const response: BaseResponse<StageCompletionData> = await apiClient.post(
    `/api/space/${spaceId}/edu/stages/complete`,
    request
  );
  return response.data!;
};

/**
 * 提交项目（作业）
 */
export const submitProject = async (
  spaceId: number,
  projectId: number
): Promise<void> => {
  await apiClient.post(`/api/space/${spaceId}/edu/projects/${projectId}/submit`);
};
